enhancements/authentication/oauth-resource-storage.md

```diff
@@ -0,0 +1,197 @@
+---
+title: Secure OAuth Resource Storage
+authors:
+  - "@sttts"
+reviewers:
+  - "@mfojtik"
+  - "@deads2k"
+  - "@stlaz"
+  - "@marun"
+  - "@polynomial"
+approvers:
+  - "@mfoktik"
+  - "anand"
+creation-date: 2020-03-30
+last-updated: 2020-05-12
+status: provisional
+see-also:
+  - "/enhancements/kube-apiserver/encrypting-data-at-datastore-layer.md"  
+replaces:
+superseded-by:
+---
+
+# Secure OAuth Resource Storage
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+1. Do we really need token read access for users. This enhancement allows us to add that with reasonable effort,
+   but we should consider also either some imperative logout subresource, an endpoint for the oauth server. This
+   will be simpler to implement, but lack the `oc get oauthaccesstokens` ability for normal users.   
+
+## Summary
+
+OAuthAccessTokens and OAuthAuthorizeToken use the object name for the security 
+sensitive token information. Hence, encrypting the value only in etcd will keep
+the token as plain text in the etcd database and in encrypted etcd backups.
+
+This enhancement is about the migration to an extended storage format of the
+two token types which adds a second secret string to the body of the resource, 
+and hence making the keys alone insensitive.
+
+The token passed to the clients will be the concatenated object name plus the new
+secret string in the body, i.e. a compound token.
+
+The authenticator in kube-apiserver will be extended to make the lookup to the
+oauth access tokens with the first half of the passed token from the client, and
+then compares the body as well before authenticating the request.
+
+## Motivation
+
+Etcd encryption without encrypted oauth access and authorize token is of limited
+value. While these tokens expire after a day usually, the token from a fresh backup
+or the output of etcdctl can be used to act as any oauth authenticated user in the 
+cluster.
+
+A side-effect of the proposed compound tokens is that it allows to hide the secret part
+in the body on LIST or GET requests without returning sensitive information. This
+way users could be allowed to read and possible delete their oauth access tokens, 
+e.g. in order to show that information via oc or console, or to logout certain tokens.
+
+### Goals
+
+- Change the `oauthaccesstokens` and `oauthauthorizetokens` storage format and schema such that
```

Comment:
how short lived are oauthaccesstokens.  I thought it was order of minutes.

Reply:
the default expiry of oauthaccesstokens is set to a day in 4.x, oauthauthorizetokens usually last for only a few seconds as they are deleted upon oauthaccesstoken retrieval